x/stake/keeper/key.go

```diff
@@ -29,169 +28,214 @@ var (
+func GetUBDsKey(delegatorAddr sdk.Address) []byte {
+	return append(UnbondingDelegationKey, delegatorAddr.Bytes()...)
 }
 
-// get the prefix keyspace for the indexs of unbonding delegations for a validator
-func GetUBDsByValIndexKey(validatorAddr sdk.Address, cdc *wire.Codec) []byte {
-	res := cdc.MustMarshalBinary(&validatorAddr)
-	return append(UnbondingDelegationByValIndexKey, res...)
+// get the prefix keyspace for the indexes of unbonding delegations for a validator
+func GetUBDsByValIndexKey(validatorAddr sdk.Address) []byte {
+	return append(UnbondingDelegationByValIndexKey, validatorAddr.Bytes()...)
 }
 
 //________________________________________________________________________________
 
 // get the key for a redelegation
+// VALUE: stake/types.RedelegationKey
 func GetREDKey(delegatorAddr, validatorSrcAddr,
-	validatorDstAddr sdk.Address, cdc *wire.Codec) []byte {
+	validatorDstAddr sdk.Address) []byte {
 
-	return append(
-		GetREDsKey(delegatorAddr, cdc),
-		append(
-			validatorSrcAddr.Bytes(),
-			validatorDstAddr.Bytes()...)...,
-	)
+	return append(append(
+		GetREDsKey(delegatorAddr.Bytes()),
+		validatorSrcAddr.Bytes()...),
+		validatorDstAddr.Bytes()...)
 }
 
 // get the index-key for a redelegation, stored by source-validator-index
+// VALUE: none (key rearrangement used)
 func GetREDByValSrcIndexKey(delegatorAddr, validatorSrcAddr,
-	validatorDstAddr sdk.Address, cdc *wire.Codec) []byte {
+	validatorDstAddr sdk.Address) []byte {
 
-	return append(
-		GetREDsFromValSrcIndexKey(validatorSrcAddr, cdc),
-		append(
-			delegatorAddr.Bytes(),
-			validatorDstAddr.Bytes()...)...,
-	)
+	return append(append(
+		GetREDsFromValSrcIndexKey(validatorSrcAddr),
+		delegatorAddr.Bytes()...),
+		validatorDstAddr.Bytes()...)
 }
 
 // get the index-key for a redelegation, stored by destination-validator-index
+// VALUE: none (key rearrangement used)
 func GetREDByValDstIndexKey(delegatorAddr, validatorSrcAddr,
-	validatorDstAddr sdk.Address, cdc *wire.Codec) []byte {
+	validatorDstAddr sdk.Address) []byte {
 
-	return append(
-		GetREDsToValDstIndexKey(validatorDstAddr, cdc),
-		append(
-			delegatorAddr.Bytes(),
-			validatorSrcAddr.Bytes()...)...,
-	)
+	return append(append(
+		GetREDsToValDstIndexKey(validatorDstAddr),
+		delegatorAddr.Bytes()...),
+		validatorSrcAddr.Bytes()...)
+}
+
+// rearrange the ValSrcIndexKey to get the REDKey
+func GetREDKeyFromValSrcIndexKey(IndexKey []byte) []byte {
+	addrs := IndexKey[1:] // remove prefix bytes
+	split := len(addrs) / 3
+	if (len(addrs) % 3) != 0 {
+		panic("unexpected key length")
+	}
+	valSrcAddr := addrs[:split]
+	delAddr := addrs[split : 2*split]
+	valDstAddr := addrs[2*split:]
+
+	return GetREDKey(delAddr, valSrcAddr, valDstAddr)
+}
+
+// rearrange the ValDstIndexKey to get the REDKey
+func GetREDKeyFromValDstIndexKey(IndexKey []byte) []byte {
+	addrs := IndexKey[1:] // remove prefix bytes
+	split := len(addrs) / 3
+	if (len(addrs) % 3) != 0 {
```

Comment:
Presumably we know the exact byte length of each address - should we assert that, and avoid these calculations?